File: server/src/encoder.js
'use strict';

/**
 * Encoder and filter capability detection.
 *
 * The product targets CPU-only boxes, so libx264 is the default and the
 * reference for the sizing tables. Hardware encoders are used only when they
 * are actually present *and* the operator asked for them (or asked for 'auto',
 * in which case we prefer hardware because it frees the CPU for scaling).
 */

const { execFile } = require('child_process');
const fs = require('fs');
const os = require('os');
const config = require('./config');
const logger = require('./logger');

const log = logger.scope('encoder');

const caps = {
  probed: false,
  ffmpegVersion: null,
  encoders: new Set(),
  filters: new Set(),
  drawtext: false,
  vaapiDevice: false,
  nvidia: false,
  cores: os.cpus().length,
  cpuModel: (os.cpus()[0] || {}).model || 'unknown',
  cpuSpeedMhz: (os.cpus()[0] || {}).speed || 0,
};

function run(bin, args, timeoutMs = 10000) {
  return new Promise((resolve) => {
    execFile(bin, args, { timeout: timeoutMs, maxBuffer: 8 * 1024 * 1024 }, (err, stdout, stderr) => {
      resolve({ err, stdout: stdout || '', stderr: stderr || '' });
    });
  });
}

async function probe() {
  const version = await run(config.ffmpegPath, ['-hide_banner', '-version']);
  if (version.err) {
    log.error('ffmpeg is not usable', version.err.message);
    caps.probed = true;
    return caps;
  }
  caps.ffmpegVersion = (version.stdout.split('\n')[0] || '').trim();

  const enc = await run(config.ffmpegPath, ['-hide_banner', '-encoders']);
  for (const line of enc.stdout.split('\n')) {
    const m = /^\s*[A-Z.]{6}\s+(\S+)/.exec(line);
    if (m) caps.encoders.add(m[1]);
  }

  const filt = await run(config.ffmpegPath, ['-hide_banner', '-filters']);
  for (const line of filt.stdout.split('\n')) {
    const m = /^\s*[TSC.]{3}\s+(\S+)/.exec(line);
    if (m) caps.filters.add(m[1]);
  }
  caps.drawtext = caps.filters.has('drawtext');

  // drawtext needs an explicit font file unless fontconfig is available, and it
  // is not in every minimal image. Find one once, at boot.
  const fontCandidates = [
    process.env.LABEL_FONT_FILE,
    '/usr/share/fonts/dejavu/DejaVuSans.ttf',
    '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',
    '/usr/share/fonts/TTF/DejaVuSans.ttf',
    '/usr/share/fonts/liberation/LiberationSans-Regular.ttf',
    '/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf',
    '/usr/share/fonts/noto/NotoSans-Regular.ttf',
  ].filter(Boolean);
  caps.fontFile = fontCandidates.find((p) => {
    try {
      return fs.existsSync(p);
    } catch (_) {
      return false;
    }
  }) || null;
  if (caps.drawtext && !caps.fontFile) {
    log.warn('no font file found — cell labels will fall back to the fontconfig default and may not render');
  }

  try {
    caps.vaapiDevice = fs.existsSync(config.vaapiDevice);
  } catch (_) {
    caps.vaapiDevice = false;
  }
  try {
    caps.nvidia = fs.existsSync('/dev/nvidia0') || fs.existsSync('/dev/nvidiactl');
  } catch (_) {
    caps.nvidia = false;
  }

  caps.probed = true;
  log.info('capabilities detected', {
    ffmpeg: caps.ffmpegVersion,
    cores: caps.cores,
    cpu: caps.cpuModel,
    drawtext: caps.drawtext,
    x264: caps.encoders.has('libx264'),
    vaapi: caps.encoders.has('h264_vaapi') && caps.vaapiDevice,
    nvenc: caps.encoders.has('h264_nvenc') && caps.nvidia,
    qsv: caps.encoders.has('h264_qsv'),
  });
  return caps;
}

/** Which encoders can actually be used on this machine right now. */
function available() {
  const list = [{ id: 'x264', label: 'libx264 (CPU)', usable: caps.encoders.has('libx264'), reason: caps.encoders.has('libx264') ? '' : 'not built into ffmpeg' }];
  list.push({
    id: 'vaapi',
    label: 'VA-API (Intel/AMD GPU)',
    usable: caps.encoders.has('h264_vaapi') && caps.vaapiDevice,
    reason: !caps.encoders.has('h264_vaapi') ? 'not built into ffmpeg' : !caps.vaapiDevice ? `${config.vaapiDevice} not present in the container` : '',
  });
  list.push({
    id: 'nvenc',
    label: 'NVENC (NVIDIA GPU)',
    usable: caps.encoders.has('h264_nvenc') && caps.nvidia,
    reason: !caps.encoders.has('h264_nvenc') ? 'not built into ffmpeg' : !caps.nvidia ? 'no NVIDIA device in the container' : '',
  });
  list.push({
    id: 'qsv',
    label: 'Quick Sync (Intel)',
    usable: caps.encoders.has('h264_qsv') && caps.vaapiDevice,
    reason: !caps.encoders.has('h264_qsv') ? 'not built into ffmpeg' : !caps.vaapiDevice ? `${config.vaapiDevice} not present in the container` : '',
  });
  return list;
}

/** Resolve the configured preference to something usable, CPU as the floor. */
function resolve(requested) {
  const want = String(config.encoderOverride || requested || 'auto').toLowerCase();
  const usable = new Set(available().filter((e) => e.usable).map((e) => e.id));

  if (want !== 'auto') {
    if (usable.has(want)) return want;
    log.warn(`requested encoder "${want}" is unavailable, falling back to libx264`);
    return 'x264';
  }
  for (const candidate of ['nvenc', 'vaapi', 'qsv']) {
    if (usable.has(candidate)) return candidate;
  }
  return 'x264';
}

/**
 * Output-stage ffmpeg arguments for the resolved encoder.
 * `comp` is the composition settings object from the store.
 */
function outputArgs(kind, comp) {
  const gop = Math.max(1, Math.round((comp.gopSeconds || 2) * (comp.fps || 30)));
  const common = [
    '-r', String(comp.fps || 30),
    '-g', String(gop),
    '-keyint_min', String(gop),
    '-sc_threshold', '0',
    '-b:v', `${comp.bitrateKbps}k`,
    '-maxrate', `${comp.maxrateKbps || comp.bitrateKbps}k`,
    '-bufsize', `${comp.bufsizeKbps || comp.bitrateKbps * 2}k`,
  ];

  switch (kind) {
    case 'nvenc':
      return [
        '-c:v', 'h264_nvenc',
        '-preset', 'p1',
        '-tune', 'ull',
        '-rc', 'cbr',
        '-zerolatency', '1',
        ...common,
      ];
    case 'vaapi':
      return [
        '-c:v', 'h264_vaapi',
        '-rc_mode', 'CBR',
        '-compression_level', '1',
        ...common,
      ];
    case 'qsv':
      // No hwupload for this path: the filtergraph ends in software NV12 and
      // h264_qsv takes it directly. Uploading to a VAAPI frames context (which
      // is what -vaapi_device would give us) produces frames this encoder
      // cannot accept, and the graph fails to configure.
      return [
        '-c:v', 'h264_qsv',
        '-preset', 'veryfast',
        '-low_power', '1',
        ...common,
      ];
    case 'x264':
    default: {
      const args = [
        '-c:v', 'libx264',
        '-preset', comp.preset || 'ultrafast',
        '-tune', 'zerolatency',
        '-profile:v', 'high',
        '-pix_fmt', 'yuv420p',
        ...common,
        '-x264-params', `nal-hrd=cbr:keyint=${gop}:min-keyint=${gop}:scenecut=0`,
      ];
      if (comp.threads && comp.threads > 0) args.push('-threads', String(comp.threads));
      return args;
    }
  }
}

module.exports = { probe, caps, available, resolve, outputArgs };

File: go-service/internal/encoder/encoder.go
// Package encoder detects which video encoders are actually usable on this
// machine and builds the ffmpeg output-stage arguments for whichever one
// gets picked. Ported from server/src/encoder.js — same detection strategy
// (ffmpeg's own -encoders list, a device file check, then a real two-frame
// test encode, since "ffmpeg lists the encoder and /dev/dri exists" is not
// the same as "this actually works": the VA-API driver is a separate
// package from ffmpeg, and older Intel GPUs need the legacy i965 driver
// rather than the modern iHD one). NVENC is deliberately not ported: the
// Rails-side ChannelComposition#ENCODERS list only offers
// auto/software/vaapi/qsv, so there is nothing that would ever request it.
package encoder

import (
	"context"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Caps is what this machine can actually do, established once at startup
// by Probe.
type Caps struct {
	Probed        bool
	FFmpegVersion string
	Encoders      map[string]bool // ffmpeg -encoders' own ids, e.g. "libx264", "h264_vaapi"
	Drawtext      bool
	FontFile      string                      // "" if none of the usual container font paths exist
	VAAPIDevice   bool                        // whether the configured device file exists
	Functional    map[string]FunctionalResult // "vaapi" | "qsv" -> did a real test encode work
}

// fontCandidates are the usual places a caption font sits in a minimal
// container image lacking fontconfig — drawtext needs an explicit font
// file without it. Same list as encoder.js's.
var fontCandidates = []string{
	"/usr/share/fonts/dejavu/DejaVuSans.ttf",
	"/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
	"/usr/share/fonts/TTF/DejaVuSans.ttf",
	"/usr/share/fonts/liberation/LiberationSans-Regular.ttf",
	"/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf",
	"/usr/share/fonts/noto/NotoSans-Regular.ttf",
}

type FunctionalResult struct {
	OK    bool
	Error string
}

var encoderLine = regexp.MustCompile(`^\s*[A-Z.]{6}\s+(\S+)`)
var filterLine = regexp.MustCompile(`^\s*[TSC.]{3}\s+(\S+)`)

func run(ctx context.Context, bin string, args []string, timeout time.Duration) (stdout, stderr string, err error) {
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, bin, args...)
	var outBuf, errBuf strings.Builder
	cmd.Stdout = &outBuf
	cmd.Stderr = &errBuf
	err = cmd.Run()
	return outBuf.String(), errBuf.String(), err
}

// Probe runs ffmpeg's own -encoders/-filters listings, checks for the VAAPI
// device file, and — only for an encoder that's both built in and has its
// device present — actually tries a two-frame encode, since a listed
// encoder is not always a working one.
func Probe(ctx context.Context, ffmpegPath, vaapiDevicePath string) Caps {
	caps := Caps{Encoders: map[string]bool{}, Functional: map[string]FunctionalResult{}}

	version, _, err := run(ctx, ffmpegPath, []string{"-hide_banner", "-version"}, 10*time.Second)
	if err != nil {
		caps.Probed = true
		return caps
	}
	if lines := strings.SplitN(version, "\n", 2); len(lines) > 0 {
		caps.FFmpegVersion = strings.TrimSpace(lines[0])
	}

	enc, _, _ := run(ctx, ffmpegPath, []string{"-hide_banner", "-encoders"}, 10*time.Second)
	for _, line := range strings.Split(enc, "\n") {
		if m := encoderLine.FindStringSubmatch(line); m != nil {
			caps.Encoders[m[1]] = true
		}
	}

	filt, _, _ := run(ctx, ffmpegPath, []string{"-hide_banner", "-filters"}, 10*time.Second)
	for _, line := range strings.Split(filt, "\n") {
		if m := filterLine.FindStringSubmatch(line); m != nil && m[1] == "drawtext" {
			caps.Drawtext = true
		}
	}

	candidates := fontCandidates
	if override := os.Getenv("LABEL_FONT_FILE"); override != "" {
		candidates = append([]string{override}, candidates...)
	}
	for _, path := range candidates {
		if _, statErr := os.Stat(path); statErr == nil {
			caps.FontFile = path
			break
		}
	}

	if vaapiDevicePath != "" {
		if _, statErr := os.Stat(vaapiDevicePath); statErr == nil {
			caps.VAAPIDevice = true
		}
	}

	for _, kind := range []string{"vaapi", "qsv"} {
		present := caps.encoderID(kind) != "" && caps.Encoders[caps.encoderID(kind)] && caps.VAAPIDevice
		if !present {
			continue
		}
		caps.Functional[kind] = testEncoder(ctx, ffmpegPath, vaapiDevicePath, kind)
	}

	caps.Probed = true
	return caps
}

func (c Caps) encoderID(kind string) string {
	switch kind {
	case "vaapi":
		return "h264_vaapi"
	case "qsv":
		return "h264_qsv"
	case "software":
		return "libx264"
	default:
		return ""
	}
}

// testEncoder actually encodes two frames — see the package comment for
// why a listed encoder isn't necessarily a working one.
func testEncoder(ctx context.Context, ffmpegPath, vaapiDevicePath, kind string) FunctionalResult {
	source := []string{"-f", "lavfi", "-i", "testsrc2=size=320x240:rate=5", "-frames:v", "2"}
	var args []string
	switch kind {
	case "vaapi":
		args = append([]string{"-hide_banner", "-loglevel", "error", "-vaapi_device", vaapiDevicePath}, source...)
		args = append(args, "-vf", "format=nv12,hwupload", "-c:v", "h264_vaapi", "-f", "null", "-")
	case "qsv":
		args = append([]string{"-hide_banner", "-loglevel", "error"}, source...)
		args = append(args, "-vf", "format=nv12", "-c:v", "h264_qsv", "-f", "null", "-")
	default:
		return FunctionalResult{OK: true}
	}
	_, stderr, err := run(ctx, ffmpegPath, args, 20*time.Second)
	if err == nil {
		return FunctionalResult{OK: true}
	}
	lines := strings.Split(strings.TrimSpace(stderr), "\n")
	last := ""
	if len(lines) > 0 {
		last = lines[len(lines)-1]
	}
	if len(last) > 160 {
		last = last[:160]
	}
	return FunctionalResult{OK: false, Error: last}
}

// Usable reports whether the given encoder id (software|vaapi|qsv) can
// actually be used right now.
func (c Caps) Usable(id string) bool {
	if id == "software" {
		return c.Encoders["libx264"]
	}
	if !c.Encoders[c.encoderID(id)] || !c.VAAPIDevice {
		return false
	}
	r, tested := c.Functional[id]
	return tested && r.OK
}

// Resolve turns a requested encoder ("auto"|"software"|"vaapi"|"qsv" — see
// ChannelComposition::ENCODERS) into the one that will actually be used,
// software as the floor. "auto" prefers hardware, since that's what frees
// the CPU for scaling multiple sources at once.
func Resolve(requested string, caps Caps) string {
	want := strings.ToLower(strings.TrimSpace(requested))
	if want == "" {
		want = "auto"
	}
	if want != "auto" {
		if caps.Usable(want) {
			return want
		}
		return "software"
	}
	for _, candidate := range []string{"vaapi", "qsv"} {
		if caps.Usable(candidate) {
			return candidate
		}
	}
	return "software"
}

// EncodeOptions is the subset of a ChannelComposition the output stage
// needs — fps/bitrate/preset, not source/layout/caption details.
type EncodeOptions struct {
	FPS         int
	BitrateKbps int
	MaxrateKbps int // 0 -> defaults to BitrateKbps
	BufsizeKbps int // 0 -> defaults to BitrateKbps * 2
	Preset      string
	GOPSeconds  float64 // 0 -> defaults to 2
}

// OutputArgs is the encoder-specific tail of the ffmpeg command: rate
// control, GOP, and the codec itself. Ported from encoder.js's
// outputArgs().
func OutputArgs(kind string, o EncodeOptions) []string {
	fps := o.FPS
	if fps <= 0 {
		fps = 30
	}
	gopSeconds := o.GOPSeconds
	if gopSeconds <= 0 {
		gopSeconds = 2
	}
	gop := int(gopSeconds*float64(fps) + 0.5)
	if gop < 1 {
		gop = 1
	}
	maxrate := o.MaxrateKbps
	if maxrate <= 0 {
		maxrate = o.BitrateKbps
	}
	bufsize := o.BufsizeKbps
	if bufsize <= 0 {
		bufsize = o.BitrateKbps * 2
	}
	common := []string{
		"-r", strconv.Itoa(fps),
		"-g", strconv.Itoa(gop),
		"-keyint_min", strconv.Itoa(gop),
		"-sc_threshold", "0",
		"-b:v", strconv.Itoa(o.BitrateKbps) + "k",
		"-maxrate", strconv.Itoa(maxrate) + "k",
		"-bufsize", strconv.Itoa(bufsize) + "k",
	}

	switch kind {
	case "vaapi":
		return append([]string{"-c:v", "h264_vaapi", "-rc_mode", "CBR", "-compression_level", "1"}, common...)
	case "qsv":
		// No hwupload here: the filtergraph ends in software NV12 and
		// h264_qsv takes it directly. Uploading to a VAAPI frames context
		// (what -vaapi_device would give it) produces frames this encoder
		// cannot accept, and the graph fails to configure.
		return append([]string{"-c:v", "h264_qsv", "-preset", "veryfast", "-low_power", "1"}, common...)
	default: // software
		preset := o.Preset
		if preset == "" {
			preset = "veryfast"
		}
		args := []string{"-c:v", "libx264", "-preset", preset, "-tune", "zerolatency", "-profile:v", "high", "-pix_fmt", "yuv420p"}
		args = append(args, common...)
		args = append(args, "-x264-params", "nal-hrd=cbr:keyint="+strconv.Itoa(gop)+":min-keyint="+strconv.Itoa(gop)+":scenecut=0")
		return args
	}
}
